File: t-tasks/src/main/java/com/teo/ttasks/ui/views/TaskDateView.java
package com.teo.ttasks.ui.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;

import java.util.Date;

import static com.teo.ttasks.util.DateUtils.sdfDayName;
import static com.teo.ttasks.util.DateUtils.sdfDayNumber;

public class TaskDateView extends View {

    private static final int dayNumberSizeSp = 24;
    private static final int dayNameSizeSp = 14;

    private String dayNumber = null;
    private String dayName = null;

    private Paint dayNumberPaint;
    private Paint dayNamePaint;

    private Rect dayNumberBounds;
    private Rect dayNameBounds;

    public TaskDateView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (dayNumber != null)
            drawDayNumber(canvas);
        if (dayName != null)
            drawDayName(canvas);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(measureWidth(widthMeasureSpec), measureHeight(heightMeasureSpec));
    }

    private void init(Context context) {
        TypedValue typedValue = new TypedValue();
        final TypedArray typedArray = context.getTheme().obtainStyledAttributes(typedValue.data, new int[]{
                android.R.attr.textColorPrimary, android.R.attr.textColorTertiary});
        final int primaryColor = typedArray.getColor(0, Color.BLACK);
        final int tertiaryColor = typedArray.getColor(1, Color.BLACK);
        typedArray.recycle();

        dayNumberPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        dayNumberPaint.setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, dayNumberSizeSp, getResources().getDisplayMetrics()));
        dayNumberPaint.setColor(primaryColor);
        dayNumberPaint.setTextAlign(Paint.Align.CENTER);
        dayNumberPaint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.NORMAL));

        dayNamePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        dayNamePaint.setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, dayNameSizeSp, getResources().getDisplayMetrics()));
        dayNamePaint.setColor(tertiaryColor);
        dayNamePaint.setTextAlign(Paint.Align.CENTER);
        dayNamePaint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.NORMAL));

        String number = "33";
        dayNumberBounds = new Rect();
        dayNumberPaint.getTextBounds(number, 0, number.length(), dayNumberBounds);

        String name = "MMM";
        dayNameBounds = new Rect();
        dayNamePaint.getTextBounds(name, 0, name.length(), dayNameBounds);
    }

    private int measureHeight(int measureSpec) {
        int size = getPaddingTop() + getPaddingBottom();
        size += dayNumberBounds.bottom + dayNumberBounds.height();
        size += dayNameBounds.bottom + dayNameBounds.height();
        size += TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, getResources().getDisplayMetrics());
        return resolveSizeAndState(size, measureSpec, 0);
    }

    private int measureWidth(int measureSpec) {
        int size = getPaddingLeft() + getPaddingRight();
        size += Math.max(dayNumberBounds.left + dayNumberBounds.width(), dayNameBounds.left + dayNameBounds.width());
        return resolveSizeAndState(size, measureSpec, 0);
    }

    public void setDate(@Nullable Date date) {
        if (date == null) {
            dayName = null;
            dayNumber = null;
        } else {
            dayName = sdfDayName.format(date);
            dayNumber = sdfDayNumber.format(date);
        }
        invalidate();
    }

    private void drawDayNumber(Canvas canvas) {
        float x = getPaddingLeft() + getWidth() / 2;
        //the y coordinate marks the bottom of the text, so we need to factor in the height
        float y = getPaddingTop() + dayNumberBounds.bottom + dayNumberBounds.height();
        canvas.drawText(dayNumber, x, y, dayNumberPaint);
    }

    private void drawDayName(Canvas canvas) {
        float x = getPaddingLeft() + getWidth() / 2;
        //the y coordinate marks the bottom of the text, so we need to factor in the height
        float y = getPaddingTop() + dayNumberBounds.bottom + dayNumberBounds.height();
        y += TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 4, getResources().getDisplayMetrics());
        y += dayNameBounds.bottom + dayNameBounds.height();
        canvas.drawText(dayName, x, y, dayNamePaint);
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/ui/views/TaskDateView.kt
package com.teo.ttasks.ui.views

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.LinearLayout
import com.teo.ttasks.R
import com.teo.ttasks.util.DateUtils
import kotlinx.android.synthetic.main.view_task_date.view.*
import org.threeten.bp.ZonedDateTime

/**
 * See [R.styleable.TaskDateView]
 *
 * @attr ref R.styleable.TaskDateView_dueDate
 */
class TaskDateView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : LinearLayout(context, attrs, defStyleAttr, defStyleRes) {

    /**
     * The date shown in this view. Can be a due date or a completion date.
     */
    var date: ZonedDateTime? = null
        set(value) {
            val equals = field == value
            if (value == null) {
                dayOfWeek.text = null
                dayOfMonth.text = null
            } else {
                dayOfWeek.text = value.format(DateUtils.formatterDayName)
                dayOfMonth.text = value.format(DateUtils.formatterDayNumber)
            }
            field = value
            if (!equals) {
                invalidate()
                requestLayout()
            }
        }

    init {
        LayoutInflater.from(context).inflate(R.layout.view_task_date, this, true)
        orientation = VERTICAL
        gravity = Gravity.CENTER_HORIZONTAL

        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.TaskDateView, 0, 0)
        try {
            date = a.getString(R.styleable.TaskDateView_date)?.let { ZonedDateTime.parse(it) }
        } finally {
            a.recycle()
        }
    }
}
